### PR TITLE
Create p:f:ConstructionData by a selected group of processes

### PR DESCRIPTION
**Pull request by peterrum**
This PR is part of the effort to introduce the new `parallel::fullydistributed::Triangulation` (see #8558). This triangulation will be optionally created in two steps: 
1. create the `p:f:ConstructionData` by a set of master processes and 
2. distributed it to the other processes. 

The PR actually adds the function `parallel::fullydistributed::Utilities::create_construction_data_from_triangulation`, which uses the serialization functionalities introduced by PR #8746 and #8748.

### CODE REVIEW

File: source/distributed/fully_distributed_tria_util.cc [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+          }
+        else
+          {
+            // 3a) recv packed ConstructionData from group-root process
+            //     (counter-part of 3c of root process)
+            MPI_Status status;
+            MPI_Probe(group_root, 0, comm, &status);
+            int l;
+            MPI_Get_count(&status, MPI_CHAR, &l);
+            char *buf = new char[l];
```
Comment (peterrum) [inline source/distributed/fully_distributed_tria_util.cc]: I have updated this section:
```cpp
            std::vector<char> buf(l);
            MPI_Recv(buf.data(), l, MPI_CHAR, group_root, 0, comm, &status);
            std::string serial_str(buf.data(), l);
```

**Comment by peterrum** [comment]
I would like to point out the difference between `parallel::fullydistributed::Utilities::create_construction_data_from_triangulation` and `parallel::fullydistributed::Utilities::create_construction_data_from_triangulation_in_goups`. While in the case of the first, the user creates the triangulation and partitions it:
```cpp
  Triangulation<dim> basetria(
    Triangulation<dim>::limit_level_difference_at_vertices);
  GridGenerator::hyper_L(basetria);
  basetria.refine_global(n_refinements);

  GridTools::partition_triangulation_zorder(
    Utilities::MPI::n_mpi_processes(comm), basetria);
  GridTools::partition_multigrid_levels(basetria);

  auto construction_data_1 = parallel::fullydistributed::Utilities::
    create_construction_data_from_triangulation(basetria, comm, true);
```
in the second case, the user provides a `std::function`/`lambda` which does the same steps:
```cpp
  auto construction_data_2 = parallel::fullydistributed::Utilities::
    create_construction_data_from_triangulation_in_goups<dim, spacedim>(
      [n_refinements](dealii::Triangulation<dim, spacedim> &tria) {
        GridGenerator::hyper_L(tria);
        tria.refine_global(n_refinements);
      },
      [](dealii::Triangulation<dim, spacedim> &tria,
         const MPI_Comm                        comm,
         const unsigned int /*group_size*/) {
        GridTools::partition_triangulation_zorder(
          Utilities::MPI::n_mpi_processes(comm), tria);
      },
      comm,
      3 /* group size */,
      true);
```
By having a function that can create/partition triangulations, we gain inside of the function the ability to decide which processes should do the work and which are idle.

Note: the example is taken from the test.

File: source/distributed/fully_distributed_tria_util.cc [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+                 other_rank < max_rank;
+                 other_rank++)
+              {
+                // create construction data for other ranks
+                auto construction_data =
+                  create_construction_data_from_triangulation(
+                    tria, comm, construct_multilevel_hierarchy, other_rank);
+                // pack
+                std::string                                       serial_str;
+                boost::iostreams::back_insert_device<std::string> inserter(
```
Comment (peterrum) [inline source/distributed/fully_distributed_tria_util.cc]: @kronbichler I guess I could use `Utilities::pack` and `Utilities::unpack` here?

File: source/distributed/fully_distributed_tria_util.cc [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+                  serial_str);
+                boost::iostreams::stream<
+                  boost::iostreams::back_insert_device<std::string>>
+                                                s(inserter);
+                boost::archive::binary_oarchive oa(s);
+                oa << construction_data;
+                s.flush();
+
+                // send construction_data
+                MPI_Send(serial_str.c_str(),
```
Comment (peterrum) [inline source/distributed/fully_distributed_tria_util.cc]: @kronbichler and a `const ierr = ` is missing here ;)

**Comment by peterrum** [comment]
Ready to be reviewed! CI will fail since the serialization of `ConstructionData` is part of another PR, which is about to be merged!

**Comment by peterrum** [comment]
I have rebased this PR. Ready to be reviewed!

File: source/distributed/fully_distributed_tria_util.cc [hunk 4]
```diff
@@ -512,6 +514,97 @@ namespace parallel
...
+
+            // Step 2: partition active cells and ...
+            serial_grid_partitioner(tria, comm, group_size);
+
+            // ... cells on the levels if multigrid is required
+            if (construct_multilevel_hierarchy)
+              GridTools::partition_multigrid_levels(tria);
+
+
+            const unsigned int max_rank_in_group =
```
Comment (kronbichler) [inline source/distributed/fully_distributed_tria_util.cc]:
```suggestion
            const unsigned int end_group =
```
You use this as the upper end of a half-open interval, whereas `max_rank_in_group` would suggest that it is still included.

File: tests/fullydistributed_grids/copy_serial_tria_grouped_01.cc [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+          Utilities::MPI::n_mpi_processes(comm), tria);
+      },
+      comm,
+      3 /* group size */,
+      true);
+
+  // 3) the result has to be identical
+  AssertThrow((construction_data_1 == construction_data_2 &&
+               construction_data_1.comm == construction_data_2.comm),
+              ExcMessage("ConstructionDatas are not the same!"));
```
Comment (kronbichler) [inline tests/fullydistributed_grids/copy_serial_tria_grouped_01.cc]: Do you want to print some information of this construction data to the output stream?

**Comment by peterrum** [comment]
@kronbichler I have implemented your suggestions! Thanks!

**Pull request merged**